File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/Auxiliary/AuxLoader/AuxTextureLoader.cpp
#include "MyProject.h"
#include "AuxTextureLoader.h"
#include "Engine/Texture.h"	// UTexture
#include "TextureInsRes.h"
#include "Prequisites.h"
#include "MClassFactory.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(AuxTextureLoader, AuxLoaderBase)

AuxTextureLoader::AuxTextureLoader()
	: Super()
{
    this->mTextureRes = nullptr;
    this->mTexture = nullptr;
}

void AuxTextureLoader::dispose()
{
	Super::dispose();
}

UTexture* AuxTextureLoader::getTexture()
{
    return this->mTexture;
}

std::string AuxTextureLoader::getOrigPath()
{
    if (this->mTextureRes != nullptr)
    {
        return this->mTextureRes->getOrigPath();
    }

    return this->mPath;
}

void AuxTextureLoader::syncLoad(
    std::string path,
	EventDispatchDelegate evtHandle,
	EventDispatchDelegate progressHandle
    )
{
    Super::syncLoad(
        path,
        evtHandle,
        progressHandle
        );

    if (this->isInvalid())
    {
        this->onStartLoad();

        this->mTextureRes = GTextureInsResMgr->getAndSyncLoadRes(
            path, 
			EventDispatchDelegate(),
			EventDispatchDelegate(),
            this->mResLoadPriority
            );
        this->onTextureLoaded((uint)0, this->mTextureRes);
    }
    else if (this->hasLoadEnd())
    {
        //this->onTextureLoaded(this->mTextureRes);
        this->onTextureLoaded((uint)0, nullptr);
    }
}

// 异步加载对象
void AuxTextureLoader::asyncLoad(
    std::string path,
	EventDispatchDelegate evtHandle,
	EventDispatchDelegate progressHandle
    )
{
	Super::asyncLoad(
		path, 
		evtHandle, 
		progressHandle
	);

    if (this->isInvalid())
    {
        this->onStartLoad();

        if (progressHandle.empty())
        {
            this->mTextureRes = GTextureInsResMgr->getAndAsyncLoadRes(
                path,
				MakeEventDispatchDelegate(
					this,
					&AuxTextureLoader::onTextureLoaded, 
					(uint)0
				)
            );
        }
        else
        {
            this->mTextureRes = GTextureInsResMgr->getAndAsyncLoadRes(
                path,
				MakeEventDispatchDelegate(
					this,
					&AuxTextureLoader::onTextureLoaded, 
					(uint)0
				),
				MakeEventDispatchDelegate(
					this,
					&AuxLoaderBase::onProgressEventHandle, 
					(uint)0
				)
            );
        }
    }
    else if (this->hasLoadEnd())
    {
        this->onTextureLoaded((uint)0, nullptr);
    }
}

void AuxTextureLoader::onTextureLoaded(uint eventId, IDispatchObject* dispObj)
{
    if (nullptr != dispObj)
    {
        this->mTextureRes = (TextureInsRes*)dispObj;

        if (this->mTextureRes->hasSuccessLoaded())
        {
            this->onLoaded();

            this->mTexture = this->mTextureRes->getTexture();
        }
        else if (this->mTextureRes->hasFailed())
        {
            this->onFailed();

			GTextureInsResMgr->unload(
                this->mTextureRes->getResUniqueId(),
				MakeEventDispatchDelegate(
					this,
					&AuxTextureLoader::onTextureLoaded, 
					(uint)0
				)
			);

            this->mTextureRes = nullptr;
        }
    }

    if (nullptr != this->mResEventDispatch)
    {
        this->mResEventDispatch->dispatchEvent(this);
    }
}

void AuxTextureLoader::unload()
{
    if(nullptr != this->mTextureRes)
    {
		GTextureInsResMgr->unload(
            this->mTextureRes->getResUniqueId(),
			MakeEventDispatchDelegate(
				this,
				&AuxTextureLoader::onTextureLoaded, 
				(uint)0
			)
        );

        this->mTextureRes = nullptr;
    }

	Super::unload();
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Auxiliary/AuxLoader/AuxTextureLoader.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");
MLoader("MyLua.Libs.Auxiliary.AuxLoader.AuxLoaderBase");

-- 纹理加载器
local M = GlobalNS.Class(GlobalNS.AuxLoaderBase);
M.clsName = "AuxTextureLoader";
GlobalNS[M.clsName] = M;

function M:ctor()
	self.mNativeTextureLoader = nil;
end

function M:dtor()
    self:dispose();
end

function M:dispose()
    if(self.mNativeTextureLoader ~= nil) then
        self.mNativeTextureLoader:dispose();
        self.mNativeTextureLoader = nil;
    end
	
	M.super.dispose(self);
end

function M:getNativeLoader()
	return self.mNativeTextureLoader;
end

function M:getTexture()
	return self.mNativeTextureLoader:getTexture();
end

function M:syncLoad(path, pThis, handle, progressHandle)
	M.super.asyncLoad(self, path, pThis, handle, progressHandle);
	
	if (self:isInvalid()) then
		if(nil == self.mNativeTextureLoader) then
			self.mNativeTextureLoader = GlobalNS.CSSystem.AuxTextureLoader.New("");
		end
		
		if(nil == progressHandle) then
			self.mNativeTextureLoader:syncLoad(path, self, self.onTextureLoaded, nil);
		else
			self.mNativeTextureLoader:syncLoad(path, self, self.onTextureLoaded, self.onProgressEventHandle);
		end
	elseif (self:hasLoadEnd()) then
		self:onBytesLoaded(nil);
	end
end

function M:asyncLoad(path, pThis, handle, progressHandle)
	M.super.asyncLoad(self, path, pThis, handle, progressHandle);
	
	if (self:isInvalid()) then
		if(nil == self.mNativeTextureLoader) then
			self.mNativeTextureLoader = GlobalNS.CSSystem.AuxTextureLoader.New("");
		end
		
		if(nil == progressHandle) then
			self.mNativeTextureLoader:asyncLoad(path, self, self.onTextureLoaded, nil);
		else
			self.mNativeTextureLoader:asyncLoad(path, self, self.onTextureLoaded, self.onProgressEventHandle);
		end
	elseif (self:hasLoadEnd()) then
		self:onTexLoaded(nil);
	end
end

function M:onTextureLoaded(dispObj)
	self.mNativeTextureLoader = dispObj[0];
    self.mEvtHandle:dispatchEvent(self);
end

return M;